prow/plugins/approve/approvers2/owners.go

```diff
@@ -0,0 +1,1024 @@
+	for file, approvers := range approvers {
+		for approver := range approvers {
+			if _, ok := approverFiles[approver]; ok {
+				approverFiles[approver].Insert(file)
+			} else {
+				approverFiles[approver] = sets.NewString(file)
+			}
+		}
+	}
+	return approverFiles
+}
+
+func findMostCoveringApprover(allApprovers []string, reverseMap map[string]sets.String, unapproved sets.String) string {
+	maxCovered := 0
+	var bestPerson string
+	for _, approver := range allApprovers {
+		filesCanApprove := reverseMap[approver]
+		if filesCanApprove.Intersection(unapproved).Len() > maxCovered {
+			maxCovered = len(filesCanApprove)
+			bestPerson = approver
+		}
+	}
+	return bestPerson
+}
+
+// temporaryUnapprovedFiles returns the list of files that wouldn't be
+// approved by the given set of approvers.
+func (o Owners) temporaryUnapprovedFiles(approvals []Approval) sets.String {
+	ap := NewApprovers(o)
+	for _, aprvl := range approvals {
+		for _, info := range aprvl.Infos {
+			ap.AddApprover(aprvl.Login, info.Reference, info.Path)
+		}
+	}
+	return ap.UnapprovedFiles()
+}
+
+// KeepCoveringApprovers finds who we should keep as suggested approvers given a pre-selection
+// knownApprovers must be a subset of potentialApprovers.
+func (o Owners) KeepCoveringApprovers(reverseMap map[string]sets.String, knownApprovals []Approval, potentialApprovers []string) sets.String {
+	if len(potentialApprovers) == 0 {
+		o.log.Debug("No potential approvers exist to filter for relevance. Does this repo have OWNERS files?")
+	}
+	keptApprovers := sets.NewString()
+
+	unapproved := o.temporaryUnapprovedFiles(knownApprovals)
+
+	for _, suggestedApprover := range o.GetSuggestedApprovers(reverseMap, potentialApprovers).List() {
+		if reverseMap[suggestedApprover].Intersection(unapproved).Len() != 0 {
+			keptApprovers.Insert(suggestedApprover)
+		}
+	}
+
+	return keptApprovers
+}
+
+// GetSuggestedApprovers solves the exact cover problem, finding an approver capable of
+// approving every OWNERS file in the PR
+func (o Owners) GetSuggestedApprovers(reverseMap map[string]sets.String, potentialApprovers []string) sets.String {
+	ap := NewApprovers(o)
+	for !ap.RequirementsMet() {
+		newApprover := findMostCoveringApprover(potentialApprovers, reverseMap, ap.UnapprovedFiles())
+		if newApprover == "" {
+			o.log.Warnf("Couldn't find/suggest approvers for each file. Unapproved: %q", ap.UnapprovedFiles().List())
+			return ap.GetCurrentApproversSet()
+		}
+		ap.AddApprover(newApprover, "", "")
+	}
+
+	return ap.GetCurrentApproversSet()
+}
+
+// GetOwnersSet returns a set containing all the Owners files necessary to get the PR approved
+func (o Owners) GetOwnersSet() sets.String {
+	owners := sets.NewString()
+	for _, fn := range o.filenames {
+		ownersFile := o.repo.FindApproverOwnersForFile(fn)
+		// If the ownersfile for fn is in the parent folder and has AllowFolderCreation enabled, we purge
+		// the file from our filenames list, because it doesn't need approval
```

Comment:
I think the `IsAutoApproveUnownedSubfolders` commit is incomplete

The approvers2 `GetOwnersSet` implementation isn't doing what the comment says.  I think it should. In other words this should be a copy-paste of https://github.com/kubernetes/test-infra/blob/f24733b609a6d2b8ce9393a8a1eaaf88d7bbb592/prow/plugins/approve/approvers/owners.go#L183-L202